pkg/dra/claims.go

```diff
@@ -0,0 +1,195 @@
+// returns an error.
+func GetResourceRequestsForResourceClaimTemplates(
+	ctx context.Context,
+	cl client.Client,
+	wl *kueue.Workload,
+	lookup func(dc corev1.ResourceName) (corev1.ResourceName, bool),
+) (map[kueue.PodSetReference]corev1.ResourceList, error) {
+	perPodSet := make(map[kueue.PodSetReference]corev1.ResourceList)
+	for i := range wl.Spec.PodSets {
+		ps := &wl.Spec.PodSets[i]
+		aggregated := corev1.ResourceList{}
+
+		// Resolve every ResourceClaim reference in the PodSet template.
+		for _, prc := range ps.Template.Spec.ResourceClaims {
+			if prc.ResourceClaimTemplateName == nil {
+				continue
+			}
+			spec, err := getClaimSpec(ctx, cl, wl.Namespace, prc)
+			if err != nil {
+				return nil, fmt.Errorf("failed to get claim spec for ResourceClaimTemplate %s in workload %s podset %s: %w", *prc.ResourceClaimTemplateName, wl.Name, ps.Name, fmt.Errorf("%w: %v", ErrClaimSpecNotFound, err))
+			}
+			if spec == nil {
+				continue
+			}
+
+			for dc, qty := range countDevicesPerClass(spec) {
+				logical, found := lookup(dc)
+				if !found {
+					return nil, fmt.Errorf("DeviceClass %s is not mapped in DynamicResourceAllocationConfig for workload %s podset %s: %w", dc, wl.Name, ps.Name, ErrDeviceClassNotMapped)
+				}
+				aggregated = utilresource.MergeResourceListKeepSum(aggregated, corev1.ResourceList{logical: qty})
+			}
+		}
+
+		if len(aggregated) > 0 {
+			perPodSet[ps.Name] = aggregated
+		}
+	}
+
+	return perPodSet, nil
+}
+
+// GetResourceRequestsForResourceClaims walks all ResourceClaims referenced by each PodSet of the Workload,
+// converts DeviceClass counts into logical resources using the provided lookup function and
+// returns the aggregated quantities per PodSet.
+//
+// If at least one DeviceClass is not present in the DynamicResourceAllocationConfig the function
+// returns an error.
+func GetResourceRequestsForResourceClaims(ctx context.Context,
+	cl client.Client,
+	wl *kueue.Workload,
+	cqName string,
+	lookup func(dc corev1.ResourceName) (corev1.ResourceName, bool)) (map[kueue.PodSetReference]corev1.ResourceList, error) {
+	log := ctrl.LoggerFrom(ctx)
+
+	perPodSet := make(map[kueue.PodSetReference]corev1.ResourceList)
+	for i := range wl.Spec.PodSets {
+		ps := &wl.Spec.PodSets[i]
+		aggregated := corev1.ResourceList{}
+
+		for _, prc := range ps.Template.Spec.ResourceClaims {
+			if prc.ResourceClaimName == nil {
+				continue
+			}
+			spec, err := getClaimSpec(ctx, cl, wl.Namespace, prc)
+			if err != nil {
+				return nil, fmt.Errorf("failed to get claim spec for ResourceClaim %s/%s in workload %s podset %s: %w", wl.Namespace, *prc.ResourceClaimName, wl.Name, ps.Name, fmt.Errorf("%w: %v", ErrClaimSpecNotFound, err))
+			}
+			if spec == nil {
+				continue
+			}
+			claimKey := ClaimKey(fmt.Sprintf("%s/%s", wl.Namespace, *prc.ResourceClaimName))
+
+			allowed := claims.addAndCheckWorkLoad(claimKey, wl.Name)
```

Comment:
addAndCheckWorkLoad adds a key to the global claims map. I never see a removal, thus perhaps leading to a memory leak. Did I miss the deletion in claimRefTracker within this PR?

Reply:
+1, good catch. This needs to be addressed